Add doctests in all functions in basic_string.py

KelvinPuyam opened the pull request:
### Describe your change:



* [ ] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [x] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [ ] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

KelvinPuyam commented:
Requesting review: @cclauss

cclauss commented on genetic_algorithm/basic_string.py:
```diff
@@ -33,15 +33,25 @@ def evaluate(item: str, main_target: str) -> tuple[str, float]:
+    Mutate a random gene of a child with another one from the list.
+    >>> random.seed(42)
+    >>> mutate("123456", list("ABCDEF"))
+    '123456'
```
Please add a mutation that changes (mutates) the list.

cclauss commented on genetic_algorithm/basic_string.py:
```diff
@@ -54,19 +64,27 @@ def select(
-        pop.append(mutate(child_1, genes))
-        pop.append(mutate(child_2, genes))
-    return pop
+        # Randomly select the second parent from the population.
```
Please revert.  Please either add tests or change the algorithm but not both in a single pull request.

KelvinPuyam commented:
I reverted back to the original code. I added the doctests for all the functions. I also showed mutation in the mutate function doctest. So in conclusion this I only added the doctests. I did not change the algorithm. Pls let me know if I should make any changes.

cclauss commented on genetic_algorithm/basic_string.py:
```diff
@@ -54,7 +64,25 @@ def select(
+    """
+    Select the second parent and generate new population
+
+    >>> import random
+    >>> random.seed(42)
+    >>> parent_1 = ("123456", 8.0)
+    >>> population_score = [("abcdef", 4.0), ("ghijkl", 5.0), ("mnopqr", 7.0)]
+    >>> genes = list("ABCDEF")
+    >>> pop = []
+    >>> child_n = int(parent_1[1]) + 1
+    >>> child_n = 10 if child_n >= 10 else child_n
+    >>> for _ in range(child_n):
+    ...     parent_2 = population_score[random.randint(0, len(population_score) - 1)][0]
+    ...     child_1, child_2 = crossover(parent_1[0], parent_2)
+    ...     pop.append(mutate(child_1, genes))
+    ...     pop.append(mutate(child_2, genes))
+    >>> len(pop) == (int(parent_1[1]) + 1) * 2
```
```suggestion
    >>> random.seed(42)
    >>> parent_1 = ("123456", 8.0)
    >>> population_score = [("abcdef", 4.0), ("ghijkl", 5.0), ("mnopqr", 7.0)]
    >>> genes = list("ABCDEF")
    >>> child_n = int(min(parent_1[1]) + 1, 10))
    >>> population = []
    >>> for _ in range(child_n):
    ...     parent_2 = population_score[random.randrange(len(population_score))][0]
    ...     child_1, child_2 = crossover(parent_1[0], parent_2)
    ...     population.extend((mutate(child_1, genes), mutate(child_2, genes)))
    >>> len(population) == (int(parent_1[1]) + 1) * 2
```

cclauss reviewed (approved):
Rock and roll!!
